### PR TITLE
Allow to choose output precision in tf_echo

### PR DESCRIPTION
**Pull request by VictorLamoine**
Improves #171 

With this PR it is possible to choose the precision of the pose output by passing an argument to the node:

```bash
rosrun tf tf_echo base tool0 _precision:=12
```

I am inspecting robot poses and I need at least millimeter precision. This allows getting the precision without cluttering the default output.

Example run:
```bash
$ rosrun tf tf_echo base tool0 _precision:=8
[ INFO] [1556030892.141355409]: Precision: 8
Failure at 1556030893.154955017
Exception thrown:Could not find a connection between 'base' and 'tool0' because they are not part of the same tree.Tf has two or more unconnected trees.
The current list of frames is:
Frame base exists with parent base_link.
Frame flange exists with parent link_6.
Frame tool0 exists with parent link_6.

Failure at 1556030893.155026131
Exception thrown:Could not find a connection between 'base' and 'tool0' because they are not part of the same tree.Tf has two or more unconnected trees.
The current list of frames is:
Frame base exists with parent base_link.
Frame flange exists with parent link_6.
Frame tool0 exists with parent link_6.

At time 1556030893.30690527
- Translation: [1.62458820, -0.14177884, 0.48463830]
- Rotation: in Quaternion [0.70637963, 0.70496671, -0.05111242, 0.03791152]
            in RPY (radian) [-3.12293846, 0.12599536, 1.56997085]
            in RPY (degree) [-178.93119320, 7.21900259, 89.95270386]

```

### CODE REVIEW

File: tf/src/tf_echo.cpp [hunk 1]
```diff
@@ -91,14 +91,25 @@ int main(int argc, char ** argv)
   }
   ros::Rate rate(rate_hz);
 
+  int precision(3);
```
Comment (tfoote) [inline tf/src/tf_echo.cpp:93]: There's no need to initialize this here it's confusing to have it here and 3 lines below.

File: tf/src/tf_echo.cpp [hunk 2]
```diff
@@ -91,14 +91,25 @@ int main(int argc, char ** argv)
...
+  {
+    ros::NodeHandle p_nh("~");
+    p_nh.param("precision", precision, 3);
+    if (precision < 1)
+    {
+      std::cerr << "Precision must be > 0\n";
+      return -1;
+    }
+    ROS_INFO_STREAM("Precision: " << precision);
```
Comment (tfoote) [inline tf/src/tf_echo.cpp]: Please update this to only print if it's overridden from the default and mention that it's being overridden by the parameter.

File: tf/src/tf_echo.cpp [hunk 3]
```diff
@@ -91,14 +91,25 @@ int main(int argc, char ** argv)
   }
   ros::Rate rate(rate_hz);
 
+  int precision(3);
+  {
```
Comment (tfoote) [inline tf/src/tf_echo.cpp:95]: The braces here are not necessary.

File: tf/src/tf_echo.cpp [hunk 4]
```diff
@@ -91,14 +91,25 @@ int main(int argc, char ** argv)
...
+    ros::NodeHandle p_nh("~");
```
Comment (tfoote) [inline tf/src/tf_echo.cpp]: Since we're now potentially using the private nodehandle multiple times. Please just create it at the top and then we can reuse it in all the conditionals.

File: tf/src/tf_echo.cpp [hunk 5]
```diff
@@ -110,15 +121,15 @@ int main(int argc, char ** argv)
...
+        std::cout.precision(precision);
         std::cout.setf(std::ios::fixed,std::ios::floatfield);
         std::cout << "At time " << echo_transform.stamp_.toSec() << std::endl;
         double yaw, pitch, roll;
         echo_transform.getBasis().getRPY(roll, pitch, yaw);
         tf::Quaternion q = echo_transform.getRotation();
         tf::Vector3 v = echo_transform.getOrigin();
         std::cout << "- Translation: [" << v.getX() << ", " << v.getY() << ", " << v.getZ() << "]" << std::endl;
-        std::cout << "- Rotation: in Quaternion [" << q.getX() << ", " << q.getY() << ", " 
+        std::cout << "- Rotation: in Quaternion [" << q.getX() << ", " << q.getY() << ", "
```
Comment (tfoote) [inline tf/src/tf_echo.cpp]: Please avoid unrelated whitespace changes. They make maintenance and change auditing harder.

**Review by tfoote — CHANGES_REQUESTED** [review]
It's fine to add this here, but this is an EOL tool. The new version is in tf2_tools: https://github.com/ros/geometry2/blob/melodic-devel/tf2_tools/scripts/echo.py I would suggest trying it out for your use case in the long term, though it will need a similar extension.

I also gave some feedback on the structure inline.

File: tf/src/tf_echo.cpp (hunk 4 above)
Comment (VictorLamoine) [reply tf/src/tf_echo.cpp]: Ok!

File: tf/src/tf_echo.cpp (hunk 3 above)
Comment (VictorLamoine) [reply tf/src/tf_echo.cpp:95]: The point was RAII: after the braces closes `p_nh` is automatically destroyed.

File: tf/src/tf_echo.cpp (hunk 1 above)
Comment (VictorLamoine) [reply tf/src/tf_echo.cpp:93]: Ok

File: tf/src/tf_echo.cpp (hunk 2 above)
Comment (VictorLamoine) [reply tf/src/tf_echo.cpp]: Ok

**Comment by VictorLamoine** [comment]
I've done modifications to fix what you suggest.
I'll look at the `tf2_tools` and try to add the same functionality!

---
It's here: https://github.com/ros/geometry2/pull/377

**Review by tfoote — APPROVED** [review]
Thanks for the quick followup as well as the follow up on geometry2!
